dxp/react_rendering.md

```diff
@@ -0,0 +1,148 @@
+-   Interface name: `com.liferay.portal.template.react.renderer.ReactRenderer`
+-   Service type: `singleton`
+
+This is the low-level service to render React components.
+
+> Note that it renders React components, not any React code, so you must provide a source which exports a React component to render.
+
+It has one method:
+
+```java
+public void renderReact(
+		ComponentDescriptor componentDescriptor, Map<String, Object> data,
+		HttpServletRequest httpServletRequest, Writer writer)
+	throws IOException;
+```
+
+which receives:
+
+1. A `ComponentDescriptor` describing the source of the React component to render. More on this in the next section.
+2. A `Map` containing arbitrary data that will be injected to the React component as initial properties. Note that the data inside the map must be compatible with JavaScript, so you cannot inject any Java object you like, only primitive types, `Map`s, `Array`s, ... that can be serialized to JSON.
+3. The current `HttpServletRequest` for which the component is being rendered.
+4. A `Writer` where the rendered HTML (and optionally the JavaScript) is written to.
+
+Note that parameter 4 is not intended to be used to write the rendered HTML and JavaScript to any place you like but to pass the `PrintWriter` from the `HttpServletResponse` or the `JspWriter` from the `PageContext` you have available when invoking the method.
+
+Also, take into account that, if you don't render the JavaScript inline (see option `positionInline` of `ComponentDescriptor`), the JavaScript will be written to the bottom of the current `HttpServletResponse`, while the HTML will be inlined into the passed `Writer`.
+
+### What the ComponentDescriptor contains
+
+`ComponentDescriptor` instances are [value objects](https://en.wikipedia.org/wiki/Value_object) that describe what needs to be rendered. A `ComponentDescriptor` may have up to five fields (of which only the first one is mandatory):
+
+1. `module`: the name of the AMD module that exports the React component to be rendered. Note that you need to provide the canonical name of the module (including its version number). To avoid hardcoding any version number you may use the instance of the `com.liferay.frontend.js.loader.modules.extender.npm.NPMResolver` service tied to the OSGi bundle containing the AMD module to be rendered.
+2. `componentId`: the `componentId` that will be given to the component's JavaScript as an initial property. Note that this field is only used if the `data` map passed to `renderReact` does **NOT** contain a `componentId` property.
+3. `dependencies`: this is deprecated; don't use it 🧐.
+4. `positionInline`: a flag to render the JavaScript directly in the current position of the HTML response or defer it to the bottom of the page. By default `false`.
+5. `propsTransformer`: optional name of an AMD module exporting a function to transform the properties before they are passed to the component.
+
+### How the React component is rendered
```

Comment:
Love this section ❤️

Reply:
:shrug:

Reply:
![lovingit](https://user-images.githubusercontent.com/905006/92721502-1a2f3c00-f35e-11ea-811c-bb6a4bc7ea6d.gif)